### PR TITLE
Swap Typha's Ctrie for Google's B-Tree library. Add snapshot size gauge

### PR DESCRIPTION
**Pull request by fasaxc**
## Description

- The B-tree's `Len()` operation is `O(1)` whereas Ctrie was `O(n)` so we can add a gauge to monitor snapshot size.
- The B-tree lib supports generics.
- The B-tree lib supports mutable updates in between snapshots (should perform better under load).
- B-tree's aren't so "exotic".  The Ctrie we were using before was a really unusual datastructure that was optimised for other use cases (concurrent insert/remove); it was very hard to understand and debug when it went wrong.

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
Typha now uses a B-tree for its internal cache, which allows it to export a Prometheus metric, typha_snapshot_size, that gives the total size of its current snapshot of the Calico datastore.
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

**Review by caseydavenport — APPROVED** [review]

**Comment by marvin-tigera** [comment]
OK, I will merge the pull request when it's ready, leave the commits as is when I merge it, and leave the branch after I've merged it.

**Pull request merged**
